pkg/hostagent/hostagent.go

```diff
@@ -777,6 +801,141 @@ func forwardSSH(ctx context.Context, sshConfig *ssh.SSHConfig, port int, local,
 	return nil
 }
 
+func (a *HostAgent) watchCloudInitProgress(ctx context.Context) {
+	logrus.Debug("Starting cloud-init progress monitoring")
+
+	a.emitEvent(ctx, events.Event{
+		Status: events.Status{
+			SSHLocalPort: a.sshLocalPort,
+			CloudInitProgress: &events.CloudInitProgress{
+				Active: true,
+			},
+		},
+	})
+
+	maxRetries := 30
+	retryDelay := time.Second
+	var sshReady bool
+
+	for i := 0; i < maxRetries && !sshReady; i++ {
+		if i > 0 {
+			time.Sleep(retryDelay)
+		}
+
+		// Test SSH connectivity
+		args := a.sshConfig.Args()
+		args = append(args,
+			"-p", strconv.Itoa(a.sshLocalPort),
+			"127.0.0.1",
+			"echo 'SSH Ready'",
+		)
+
+		cmd := exec.CommandContext(ctx, a.sshConfig.Binary(), args...)
+		if err := cmd.Run(); err == nil {
+			sshReady = true
+			logrus.Debug("SSH ready for cloud-init monitoring")
+		}
+	}
+
+	if !sshReady {
+		logrus.Warn("SSH not ready for cloud-init monitoring, proceeding anyway")
+	}
+
+	args := a.sshConfig.Args()
+	args = append(args,
+		"-p", strconv.Itoa(a.sshLocalPort),
+		"127.0.0.1",
+		"sudo", "tail", "-n", "+1", "-f", "/var/log/cloud-init-output.log",
+	)
+
+	cmd := exec.CommandContext(ctx, a.sshConfig.Binary(), args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		logrus.WithError(err).Warn("Failed to create stdout pipe for cloud-init monitoring")
+		return
+	}
+
+	if err := cmd.Start(); err != nil {
+		logrus.WithError(err).Warn("Failed to start cloud-init monitoring command")
+		return
+	}
+
+	scanner := bufio.NewScanner(stdout)
+	cloudInitFinished := false
+
+	for scanner.Scan() {
```

Comment:
This loop appears to run indefinitely.
The `ssh -F … sudo tail -n +1 -f /var/log/cloud-init-output.log` process continues after `limactl start --progress` finishes.
Also, some times `limactl start --progress` never stop even if `Cloud-init v. 25.1.2-0ubuntu0~24.04.1 finished` received.

Reply:
I have tried to reproduce this but couldn't as everything works fine. Although, I think a timeout context can be added for safety purpose. Can you share how you were able to find the bug?